### PR TITLE
Add the creation of a staging project to its history

### LINKED ISSUE #7051: Staging Project History should include project creation
The history of `StagingProject` doe not include creation of the project. It should.

### PR DESCRIPTION
**Pull request by dmarcoux**
Closes #7051 

Screenshot_2019-04-01 Open Build Service

### CODE REVIEW

**Comment by dmarcoux** [comment]
[Address Feedback](https://trello.com/c/K9U6WlG1/1081-address-feedback)

**Comment by dmarcoux** [comment]
I will rebase on _master_ once #7188 is merged and adapt the code (mostly moving it to the new files)

File: src/api/app/models/staging/staging_project.rb [hunk 1]
```diff
@@ -8,6 +8,8 @@ class StagingProject < Project
     after_save :update_staging_workflow_on_backend
     after_destroy :update_staging_workflow_on_backend
     before_create :add_managers_group
+    after_create :create_project_log_entry
+    before_update :create_project_log_entry, if: proc { |staging_project| staging_project.staging_workflow_id_was.nil? }
```
Comment (bgeuken) [inline src/api/app/models/staging/staging_project.rb]: You probably could drop the `.nil?` by using `unless`. It might even be possible to shorten this to `unless: :staging_workflow_id_was`

File: src/api/app/views/webui2/webui/staging/projects/_history_elements.html.haml [hunk 2]
```diff
@@ -2,8 +2,14 @@
...
+      - when 'unstaged_request', 'staged_request'
+        = link_to("##{element.bs_request.number}", request_path(element.bs_request.number))
+        for package
+        %b= element.package_name
+        submitted by
+        %b= element.user_name
+      - when 'staging_project_created'
+        %b #{time_ago_in_words(Time.at(element.datetime.to_i))} ago
+        by
+        %b= element.user_name
```
Comment (bgeuken) [inline src/api/app/views/webui2/webui/staging/projects/_history_elements.html.haml]: That's more of a nitpick... I would prefer to move this to a helper. So we don't end up with a big switch condition in the view.

File: src/api/app/models/staging/staging_project.rb (hunk 1 above)
Comment (dmarcoux) [reply src/api/app/models/staging/staging_project.rb]: It's longer for sure, but it is also much clearer, so I prefer to leave it as it is.

File: src/api/app/views/webui2/webui/staging/projects/_history_elements.html.haml (hunk 2 above)
Comment (dmarcoux) [reply src/api/app/views/webui2/webui/staging/projects/_history_elements.html.haml]: I was also considering if I should do it or not when I wrote the code. I will move it to a helper.

Comment (dmarcoux) [reply src/api/app/views/webui2/webui/staging/projects/_history_elements.html.haml]: Done, let me know what you think.

File: src/api/app/models/staging/staging_project.rb [hunk 3]
```diff
@@ -225,5 +227,14 @@ def update_staging_workflow_on_backend
     def add_managers_group
       assign_managers_group(staging_workflow.managers_group)
     end
+
+    def create_project_log_entry
+      ProjectLogEntry.create!(
+        project: self,
+        user_name: User.current.login,
+        event_type: :staging_project_created,
```
Comment (dmarcoux) [inline src/api/app/models/staging/staging_project.rb]: This is annoying in the specs when we need to create a few staging projects, but don't necessarily want to login (when testing _index_ actions for example)... It fails because `User.current` is `nil`

Comment (coolo) [reply src/api/app/models/staging/staging_project.rb]: You can add a trait to the factory that adds skip_create. But if you also want the staging project in the backend, you better login in before.

Comment (dmarcoux) [reply src/api/app/models/staging/staging_project.rb]: If possible, I would like to avoid skipping the callbacks. Let's see what the others think.

Comment (bgeuken) [reply src/api/app/models/staging/staging_project.rb]: Solution 1. wouldn't work in other cases where `User.current` is needed. That's probably why we go with solution 2. in most of such cases.

The underlying problem is that we still have authorization code in some models. Until we solved this I would go with solution 2.

File: src/api/app/models/staging/staging_project.rb (hunk 1 above)
Comment (bgeuken) [reply src/api/app/models/staging/staging_project.rb]: IMO, as soon as someone knows that `if` / `unless` accepts symbols it's not. And then you have much more stuff to mentally parse.

**Comment by dmarcoux** [comment]
I avoided changing all the specs by switching
```
User.current.login
```
to
```
User.current_or_nobody.login
```

I'm not a fan of this, but I don't see another way beside putting 30+ `login` calls across a bunch of spec files.

**Comment by dmarcoux** [comment]
This `ProjectLogEntry` also needs to be displayd in the API, so add it after #7225 is merged.

**Comment by coolo** [comment]
I don't like accepting nobody here - IMO accepting broken code for the sake of less noisy tests is wrong.

File: src/api/app/models/concerns/staging_project.rb [hunk 4]
```diff
@@ -227,6 +229,15 @@ def add_managers_group
     assign_managers_group(staging_workflow.managers_group)
   end
 
+  def create_project_log_entry
+    ProjectLogEntry.create!(
+      project: self,
+      user_name: User.current_or_nobody.login, # User.current should be defined, but we don't want this to fail if it's somehow not
```
Comment (coolo) [inline src/api/app/models/concerns/staging_project.rb]: audit logs are pretty useless if they say nobody

**Comment by coolo** [comment]
why is the coverage so low btw? not rendering views?

**Comment by coolo** [comment]
@dmarcoux the more I think about it the more I come to this conclusion: having the model create audit logs / history is just as bad as models checking permissions. 

Basically every usage of User.current in model code is questionable. Another flag that this a bad idea is that your callback contains quite a big if.

**Comment by coolo** [comment]
(And of course I tried, but the rspec is basically impossible to write sanely - but https://github.com/coolo/open-build-service/commit/da1211d0945de2b88b86a0500e11bb5ca24b6ce5 is at the edge of acceptable :)

File: src/api/app/models/concerns/staging_project.rb (hunk 4 above)
Comment (dmarcoux) [reply src/api/app/models/concerns/staging_project.rb]: I agree... I tried it, but it's true that it's useless if they say nobody.

**Comment by dmarcoux** [comment]
> @dmarcoux the more I think about it the more I come to this conclusion: having the model create audit logs / history is just as bad as models checking permissions.
> 
> Basically every usage of User.current in model code is questionable. Another flag that this a bad idea is that your callback contains quite a big if.

I agree. I need to read more on the subject. I don't want to merge bad code or having to "pollute" the specs with a bunch of `login` calls.

**Comment by dmarcoux** [comment]
I ended up calling the method `create_project_log_entry` directly in the controller and passing the user to avoid the dependency on `User.current` in the model. I don't find this solution amazing because it's easy to forget to call the method, but at least, the model doesn't depend on `User.current` anymore and we don't need to change 50+ specs to "fix" them by calling `login` due to this dependency.

**Comment by dmarcoux** [comment]
Ping @bgeuken :slightly_smiling_face:!

As you already had a look before, I want to be sure you had the chance to do so again before merging this PR.

File: src/api/app/helpers/webui/staging/project_helper.rb [hunk 5]
```diff
@@ -6,4 +6,22 @@ def icon_for_checks(checks, missing_checks)
...
+    capture do
+      concat(link_to("##{history.bs_request.number}", request_path(history.bs_request.number)))
+      concat(' for package ')
+      concat(content_tag(:b, history.package_name))
+      concat(' submitted by ')
+      concat(content_tag(:b, history.user_name))
+    end
+  end
+
+  def staging_project_created_history(history)
```
Comment (hennevogel) [inline src/api/app/helpers/webui/staging/project_helper.rb]: Why a helper and not a partial?

Comment (coolo) [reply src/api/app/helpers/webui/staging/project_helper.rb]: helpers are much faster?

Comment (dmarcoux) [reply src/api/app/helpers/webui/staging/project_helper.rb]: They are both really small, so I think using a helper is fine in this case. I don't have a strong opinion on this, I'm also fine with using a partial so I  can change it if you think it's a blocker.

Comment (coolo) [reply src/api/app/helpers/webui/staging/project_helper.rb]: Lesson learned from refactoring repositories page: if the partial is small and repeated often, it should be a helper.

Comment (vpereira) [reply src/api/app/helpers/webui/staging/project_helper.rb]: "much faster" without benchmark is just a guess. I think the partial is easier to maintain, beside it's the "Rails way" https://apidock.com/rails/ActionView/Helpers/TextHelper/concat

**Comment by vpereira** [comment]
> @dmarcoux the more I think about it the more I come to this conclusion: having the model create audit logs / history is just as bad as models checking permissions.
> 
> Basically every usage of User.current in model code is questionable. Another flag that this a bad idea is that your callback contains quite a big if.

I think the usage User.current in model is questionable too, in a pure MVC design, however we have different ways to interact with a model, than just via webui controller. Doing that in the controller, we will have to repeat the same code to WebUI, API and Background jobs (do we have any other entry point which do we have to cover?). So if we don't want to have that in the model, i.e background job, being fired by a callback, then an [Observers](https://github.com/rails/rails-observers) would make more sense to trigger Audit than direct on the controller.

**Comment by coolo** [comment]
On 4/1/19 9:22 AM, Victor Pereira wrote:

> "much faster" without benchmark is just a guess. I think the partial is easier to maintain, beside it's the "Rails way"
> https://apidock.com/rails/ActionView/Helpers/TextHelper/concat
> 
How many more benchmarks do you have to see before you believe it?

Greetings, Stephan

**Comment by coolo** [comment]
> On 4/1/19 9:22 AM, Victor Pereira wrote: "much faster" without benchmark is just a guess. I think the partial is easier to maintain, beside it's the "Rails way" https://apidock.com/rails/ActionView/Helpers/TextHelper/concat
> How many more benchmarks do you have to see before you believe it? Greetings, Stephan

c84084c24f4e7d0b9d46dd5d961b69582a23065c

**Comment by coolo** [comment]
> I think the usage User.current in model is questionable too, in a pure MVC design, however we have different ways to interact with a model, than just via webui controller. Doing that in the controller, we will have to repeat the same code to WebUI, API and Background jobs (do we have any other entry point which do we have to cover?). So if we don't want to have that in the model, i.e background job, being fired by a callback, then an [Observers](https://github.com/rails/rails-observers) would make more sense to trigger Audit than direct on the controller.

That rails observers won't help you. They are just a different way to organize callbacks (and have been rightfully removed from rails) - you still have no User available in them. So you would need to rely on static variables just as well - which lead to the whole desaster.

What you need is an observer that is setup by controllers. The trick is to set this up without repeating code.

File: src/api/app/helpers/webui/staging/project_helper.rb (hunk 5 above)
Comment (hennevogel) [reply src/api/app/helpers/webui/staging/project_helper.rb]: > I don't have a strong opinion on this

Me neither. If it's because of performance and it helps, so be it.

File: src/api/app/controllers/webui/staging/projects_controller.rb [hunk 6]
```diff
@@ -26,6 +26,8 @@ def create
 
         if staging_project.valid? && staging_project.store
           flash[:success] = "Staging project with name = \"#{staging_project}\" was successfully created"
+          staging_project.create_project_log_entry(User.current)
```
[the remark this answers is not in the document]
Comment (hennevogel) [reply src/api/app/controllers/webui/staging/projects_controller.rb]: We already have a job `CreateProjectLogEntryJob`, but we can optimize this later.

[the remark this answers is not in the document]
Comment (dmarcoux) [reply src/api/app/controllers/webui/staging/projects_controller.rb]: I agree with you and I had a look. To properly do this, I would need more time as I would like to refactor [this](https://github.com/openSUSE/open-build-service/blob/26cfc4414ed79aaf1b13ffe8f38074eb6c0e81ca/src/api/app/jobs/create_project_log_entry_job.rb#L5) to make it reusable for all types of `ProjectLogEntry`. Right now, `ProjectLogEntry.create_from` depends on having a request, which is not the case for our case here.

I will add a card to the staging workflow backlog to not forget this.

File: src/api/app/helpers/webui/staging/project_helper.rb (hunk 5 above)
Comment (dmarcoux) [reply src/api/app/helpers/webui/staging/project_helper.rb]: I pushed changes to use a partial. Please have a look.

**Comment by dmarcoux** [comment]
Please have a look again, I'm using partials now instead of methods in a helper.

**Comment by coolo** [comment]
If the staging workflow is created, A and B are automatically created - and their creation is not recorded (talking about duplicating code).

But I wonder about something else now: will this history be filtered in any way (or cleaned up)?
Compare https://github.com/openSUSE/open-build-service/pull/6744#issuecomment-453145556

**Comment by coolo** [comment]
BTW: the history lookup deserves an order call - they are displayed in rather interesting order.

**Comment by coolo** [comment]
So as @vpereira wants to see benchmarks I went and took the 2 versions @dmarcoux coded.

The current PR code (partials):

```
Rendered webui2/webui/staging/projects/_history_elements.html.haml (418.7ms)
```

The previous version (helper)
```
Rendered webui2/webui/staging/projects/_history_elements.html.haml (227.4ms)
```

And this isn't with insane values of log entries - but with 401 entries. And without cleanup, this will be a *very* low number soonish.

File: src/api/app/controllers/webui/staging/projects_controller.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (vpereira) [reply src/api/app/controllers/webui/staging/projects_controller.rb]: great @dmarcoux!

**Comment by dmarcoux** [comment]
I didn't expect this to be such a pain...

Regarding the cleanup, I agree it should be done. However, it will have to be in another PR since it also touches the other `ProjectLogEntry` records creating in other parts of the staging workflow.

**Comment by dmarcoux** [comment]
> If the staging workflow is created, A and B are automatically created - and their creation is not recorded

Done

**Comment by dmarcoux** [comment]
I created a [card](https://trello.com/c/SeW31Qda/1673-cleanup-a-staging-project-history-when-it-is-merged) regarding the cleanup of the history

**Comment by coolo** [comment]
User.current strikes back - I like it :)
